### ISSUE TITLE
Drastically speed up `restic ls` by limiting to just a specific folder

### ISSUE THREAD

**Issue by mholt** [restic/restic#1940]
Output of `restic version`
--------------------------

```
restic 0.9.2 (v0.9.2-0-gbeb208e1) compiled with go1.10.2 on darwin/amd64
```

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

I'd be interested in modifying `restic ls` so that it can be scoped to just a specific directory/tree. For my purposes, I want to list the files in a certain folder -- and not subfolders -- without scanning the whole repository.

Scanning the whole repository takes about 5 minutes in my case, and the repo has only about 65 GB.

Scanning just the folder should (in theory) take only a couple seconds, based on my tests.

I am willing to submit a PR, of course. A little guidance would be appreciated about where to start in the code!

Did restic help you or made you happy in any way?
-------------------------------------------------

Yes, as usual. :) You already know it from me! :wink:

**Comment by fd0** [comment]
Good idea. It always bothered me that restic works different from `ls`, which is non-recursive by default. I can imagine the following user interface:
 * `restic ls latest` works as it does not, lists files recursively
 * `restic ls latest /` lists only the root directory
 * `restic ls latest /foo` lists only the contents of `/foo`
 * `restic ls latest --recursive /home/user` lists everything below `/home/user` recursively

The only use case that does not work any more is when multiple snapshot IDs are passed.

The place to start in the code is here: https://github.com/restic/restic/blob/master/cmd/restic/cmd_ls.go#L67

You would need to parse the second argument as a path, returning ignores from the function passed to `Walk()`, and only print things when the path passed to the function is a sub-path of the requested path.

Let me know if you need more hints :)

**Comment by mholt** [comment]
Sounds good! Thanks for taking the time to give feedback/hints!

A couple thoughts before I dive in:

- The Walk function -- it will still traverse the whole tree (albeit quicker since we'll immediately ignore non-matching nodes)?
- The traversal -- I think it is depth-first? This could be more efficient, at least for this particular task, if it's breadth-first. Not sure if that matters, but I figured I'd raise the question.
- The user interface -- what is the use case for listing multiple snapshots on `ls`? If we wanted to preserve that, we could always just use a specific flag for the directory filter.

**Comment by fd0** [comment]
> The Walk function -- it will still traverse the whole tree (albeit quicker since we'll immediately ignore non-matching nodes)?

It won't traverse trees for which the `WalkFunc` returns `SkipNode`, see https://godoc.org/github.com/restic/restic/internal/walker#WalkFunc

So it'll only walk the trees which are absolutely necessary (module bugs).

> The traversal -- I think it is depth-first? This could be more efficient, at least for this particular task, if it's breadth-first. Not sure if that matters, but I figured I'd raise the question.

It's indeed depth-first, and I think it will not make a difference in terms of performance at all. Please don't worry about efficiency for the first implementation, that's usually not a good strategy ;) I can say that with certainty, because I've been down this particular rabbit hole **way** too many times already.

That `Walk()` does depth-first is even better, since that's what users expect from a recursive listing, so you don't need to buffer any values, but print the nodes as they come.

> The user interface -- what is the use case for listing multiple snapshots on ls? If we wanted to preserve that, we could always just use a specific flag for the directory filter.

I honestly have no idea why that is in there -- which is a good indication that we can just remove it.

**Comment by mholt** [comment]
> It won't traverse trees for which the WalkFunc returns SkipNode, see https://godoc.org/github.com/restic/restic/internal/walker#WalkFunc

Okay, cool!

> That Walk() does depth-first is even better, since that's what users expect from a recursive listing, so you don't need to buffer any values, but print the nodes as they come.

Great points. Depth-first is fine then!

> I honestly have no idea why that is in there -- which is a good indication that we can just remove it.

Alrighty -- I'll tinker and see what seems best. Thanks!

**Comment by fd0** [comment]
Closing, this has been implemented in #1941
